smp/__init__.py

```diff
@@ -0,0 +1,577 @@
+            res = self.libc.dl_iterate_phdr(cmp_callback, data)
+            if res == 1 and found_module_name:
+                self.libmkl = CDLL(found_module_name)
+            else:
+                self.libmkl = None
+        except:
+            self.libmkl = None
+
+    def _load_omp(self):
+        try:
+            global found_module_name
+            self.libc = CDLL(libc_module_name)
+            found_module_name = None
+            CMPFUNC = CFUNCTYPE(c_int, c_void_p, c_size_t, c_char_p)
+            cmp_callback = CMPFUNC(callback)
+
+            data = c_char_p(omp_gnu_module_name.encode('utf-8'))
+            res = self.libc.dl_iterate_phdr(cmp_callback, data)
+            if res == 1 and found_module_name:
+                self.libomp = CDLL(found_module_name)
+            else:
+                data = c_char_p(omp_intel_module_name.encode('utf-8'))
+                res = self.libc.dl_iterate_phdr(cmp_callback, data)
+                if res == 1 and found_module_name:
+                    self.libomp = CDLL(found_module_name)
+                else:
+                    self.libomp = None
+        except:
+            self.libomp = None
+
+def get_native_wrapper():
+    global native_wrapper_list
+    global native_wrapper_lock
+
+    native_wrapper_lock.acquire()
+    native_wrapper = native_wrapper_list.get(os.getpid())
+    if not native_wrapper:
+        native_wrapper = NativeWrapper()
+        native_wrapper_list[os.getpid()] = native_wrapper
+    native_wrapper_lock.release()
+
+    return native_wrapper
+
+def mkl_set_num_threads(n):
+    native_wrapper = get_native_wrapper()
+    if native_wrapper.is_mkl_found():
+        native_wrapper.mkl_set_num_threads(n)
+
+def omp_set_num_threads(n):
+    native_wrapper = get_native_wrapper()
+    if native_wrapper.is_omp_found():
+        native_wrapper.omp_set_num_threads(n)
+
+def get_affinity():
+    if sched_getaffinity:
+        return sched_getaffinity(0)
+    else:
+        if PsutilProcess:
+            p = PsutilProcess()
+            return p.cpu_affinity()
+        else:
+            return [i for i in range(cpu_count())]
+
+def set_affinity(mask):
+    if sched_setaffinity:
+        sched_setaffinity(0, mask)
+    else:
+        if PsutilProcess:
+            p = PsutilProcess()
+            p.cpu_affinity(mask)
+        else:
+            if os.name == "posix":
+                omp_set_num_threads(len(mask))
+
+def set_proc_affinity(process_count, process_id):
+    if process_count == 1:
+        return
+
+    cpu_list = list(get_affinity())
+    cpu_count = len(cpu_list)
+
+    global oversubscription_factor
+    if cpu_count < oversubscription_factor:
+        oversubscription_factor = cpu_count
+
+    threads_per_process = oversubscription_factor
+    if cpu_count >= process_count:
+        threads_per_process = threads_per_process*int(round(float(cpu_count)
+                              /float(process_count)))
```

Comment:
I am not sure I understand what this block is really doing.

Generally, the entire file has only 2 comments/hints. I have not been involved in any of the discussions about this topic so it's hard to follow what exactly is happening. The description in the other file does not help much either (as it does not go beyond re-phrasing the names of the arguments). There is no explanation what so ever of what this module is supposed to do, e.g. what the different input arguments really mean and how this works.

I suggest you add a high-level description of this file is really doing and add more specific comment to each function. Things like lines 223-225 could also use some comment.